### ISSUE TITLE
Fix invalid pathfinding for AI heroes

### ISSUE THREAD

**Issue by ihhub** [ihhub/fheroes2#6969]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

Based on comment [here](https://github.com/ihhub/fheroes2/pull/6959#discussion_r1161099317).

The current implementation of pathfinding contains an idea that AI heroes can go "through" neutral monsters creating complex paths like capturing a mine and defeating monsters on the way. If neutral monsters are too powerful for a hero than tiles reachable for monsters are marked as inaccessible. This brings problem that heroes might attack these more powerful monsters based on logic in **ai_normal_hero.cpp**. However, at the end they aren't able to reach them.

We need to rework pathfinding logic for AI to avoid these cases.

### Save file

No need.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , would you be interested in fixing this issue since you've done similar tasks in the past? I'm asking to avoid double work being done.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , "transparent" neutral monsters is indeed a good idea but it has some flaws in it such as:
- fixed coefficient per map scan. It changes only when a hero has nothing to do
- cannot attack neutrals (the root cause of the problem) even if a hero is set to do so

I see as a solution to store pathfinding information for different coefficients and run through them if we are targeting neutral units. Of course it is slower but at the same time it gives flexibility. We can modify the existing search of accessible tiles with multiple coefficients and store for each tile the minimum strength coefficient which is required is order to pass there. In this case it won't impact the performance much. What do you think?

**Comment by oleg-derevenetz** [comment]
@ihhub let me think a bit more about this, maybe I'll come to some more or less elegant solution.

**Comment by ihhub** [comment]
Putting this issue as high priority as it is a blocker for any other improvements with AI on Adventure Map.

**Comment by oleg-derevenetz** [comment]
@ihhub I took a look at the current implementation, and to be honest, it's done pretty sensibly. AI pathfinder coefficient is a subject to change (via `setArmyStrengthMultiplier()` method) and AI planner (more specifically, `Normal::HeroesTurn()` method) will change it if there are no valuable targets with "better" coefficients are found. What I found suspicious is that `ObjectValidator` does not check for object's reachability in general case and this may cause issues like "I want to reach that object but there is no path to it", but this needs to be checked on specifically crafted maps where this problem would manifest itself.
